[concurrency] IRGen: emit emit_hop_to_executor instructions.

eeckstein opened the pull request:
This basically means to create a suspension point and call the swift_task_switch runtime function.

rdar://71099289

aschwaighofer commented on lib/IRGen/GenFunc.cpp:
```diff
@@ -2425,3 +2425,111 @@ IRGenFunction::createAsyncDispatchFn(const FunctionPointer &fnPtr,
   return dispatch;
 }
+
+void IRGenFunction::emitSuspensionPoint(llvm::Value *toExecutor,
```
IRGenFunction has a call to getAsyncExecutor(), currently this always returns the function's executor argument. This is not longer correct after this change.

We should change it so that it sets up the 'current executor' on function entry and that every suspension point resets it.

The call to `coro_suspend_async` returns a tuple of three arguments which are the arguments to the partial function. You can use the return tuple to set the new executor.

```
  %new_current_exectuor = extract_value %suspend_call, #executor_arg_index

DougGregor commented on include/swift/Runtime/RuntimeFunctions.def:
```diff
@@ -1506,6 +1506,15 @@ FUNCTION(TaskCreateFunc,
+// void swift_asyncSuspend(
+//     Executor *current, Executor *target, AsyncTask *task);
+FUNCTION(AsyncSuspendFunc,
+         swift_asyncSuspend, SwiftCC,
```
Can we call this `swift_task_suspend` or similar, because all of the task stuff starts with `swift_task_`.

aschwaighofer commented:
The required llvm async lowering change is in swift/main as of https://github.com/apple/llvm-project/commit/f681f671e2e95389476c68597b2324f15e09473e

rjmccall commented on lib/IRGen/GenFunc.cpp:
```diff
@@ -2433,3 +2434,83 @@ IRGenFunction::createAsyncDispatchFn(const FunctionPointer &fnPtr,
+        Builder.CreateRet(&*IGF.CurFn->arg_begin());
+      },
+      false /*isNoInline*/));
+}
```
I'm a little confused about what this function does.  It looks to be avoiding assuming stuff about the resumption function signature, but don't we assume a bunch of stuff about function signatures in the async lowering anyway?  Or have we managed to avoid that so far?

aschwaighofer replied:
An async suspend point needs to know how-to restore the current context when resuming from the suspend point (how-to get from the resume function's context argument to the current context when the continuation function is called). This is described as a function pointer argument to the async suspend intrinsic; LLVM's async lowering guarantees inlining of this function.

Another example of this: A suspension point that models an async call restores the current context by loading the parent context from the callee context (the current implementation does caller pop).

aschwaighofer replied:
> Or have we managed to avoid that so far?
Yes. LLVM's async lowering tries to make as little assumption about the signature as possible. The `llvm.coro.id.async` intrinsic describes at which argument index the context argument resides. The `llvm.coro.suspend.async` intrinsic describes how-to restore the current context from the resume function's context argument.

rjmccall commented on lib/IRGen/GenFunc.cpp:
```diff
@@ -2433,3 +2434,83 @@ IRGenFunction::createAsyncDispatchFn(const FunctionPointer &fnPtr,
+  arguments.push_back(getAsyncExecutor());
+  arguments.push_back(getAsyncContext());
+
+  emitSuspendAsyncCall(arguments);
```
Could you emit a fast-path branch around this that avoids the suspension point when the executors are exactly equal?

aschwaighofer replied:
Yes. Erik and I have discussed this as a follow-up optimization.

rjmccall reviewed (commented):
Generally looks great.

eeckstein committed:
runtime: work around the missing task-enqueue implementation

Instead of assert, just run the job immediately.
Needed to make the hop_to_executor IRGen test of the following commit working.

eeckstein committed:
IRGen: fix the LLVM struct definition of AsyncTask.

The object header was missing.

eeckstein committed:
IRGen: emit emit_hop_to_executor instructions.

This basically means to create a suspension point and call the swift_task_switch runtime function.

rdar://71099289

eeckstein commented:
@swift-ci smoke test

eeckstein commented:
@swift-ci smoke test

DougGregor commented on test/IRGen/async/run-switch-executor.swift:
```diff
@@ -0,0 +1,54 @@
+// REQUIRES: executable_test
+// REQUIRES: concurrency
+// UNSUPPORTED: use_os_stdlib
+// UNSUPPORTED: CPU=arm64e
```
This should be XFAIL, right?

eeckstein replied:
I patched the runtime to make this test work (89007e2e0400577f1d1fcadbe6ea8b75dd712a70)
